### PR TITLE
DOC: optimize: Add Examples to fmin_bfgs docstring

### PR DESCRIPTION
**Pull request by Chachay**
#### Reference issue
Picked one function listed in #7168

#### What does this implement/fix?
Add a simple example to docstring.

#### Additional information
Though scipy.optimize has the defined Rosebrock function, I defined it inline to give the idea for users.

### CODE REVIEW

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -1088,6 +1088,41 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
...
+    ...     return (1-x[0])**2 + 100*(x[1]-x[0]**2)**2
+
+    >>> x0 = np.array([-3, -4])
+    >>> fmin_bfgs(rosenbrock, x0)
+    Warning: Desired error not necessarily achieved due to precision loss.
+            Current function value: 0.000000
+            Iterations: 64
+            Function evaluations: 330
+            Gradient evaluations: 106
+    array([0.99999669, 0.99999329])
```
Comment (tylerjereddy) [inline scipy/optimize/optimize.py:1109]: Looks like the CI refguide (documentation check) failure is real because of a small difference:

```
2021-05-16T16:37:56.5086553Z scipy.optimize.fmin_bfgs
2021-05-16T16:37:56.5088080Z ------------------------
2021-05-16T16:37:56.5088624Z 
2021-05-16T16:37:56.5089732Z File "build/testenv/lib/python3.8/site-packages/scipy/optimize/optimize.py", line 1103, in fmin_bfgs
2021-05-16T16:37:56.5090575Z Failed example:
2021-05-16T16:37:56.5091325Z     fmin_bfgs(rosenbrock, x0)
2021-05-16T16:37:56.5092012Z Expected:
2021-05-16T16:37:56.5092758Z     Warning: Desired error not necessarily achieved due to precision loss.
2021-05-16T16:37:56.5093610Z             Current function value: 0.000000
2021-05-16T16:37:56.5094354Z             Iterations: 64
2021-05-16T16:37:56.5095095Z             Function evaluations: 330
2021-05-16T16:37:56.5095858Z             Gradient evaluations: 106
2021-05-16T16:37:56.5096580Z     array([0.99999669, 0.99999329])
2021-05-16T16:37:56.5097249Z Got:
2021-05-16T16:37:56.5097913Z     Optimization terminated successfully.
2021-05-16T16:37:56.5098687Z              Current function value: 0.000000
2021-05-16T16:37:56.5099437Z              Iterations: 65
2021-05-16T16:37:56.5100163Z              Function evaluations: 246
2021-05-16T16:37:56.5100920Z              Gradient evaluations: 82
2021-05-16T16:37:56.5101644Z     array([0.99999232, 0.99998463])
```

File: scipy/optimize/optimize.py [hunk 2]
```diff
@@ -1088,6 +1088,41 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
     allvecs  :  list
         The value of xopt at each iteration. Only returned if retall is True.
 
+    Examples
+    --------
+    >>> import numpy as np
```
Comment (tylerjereddy) [inline scipy/optimize/optimize.py:1093]: This import is implicit in our docstrings so can be removed (though it has caused some controversy recently because it is easier to copy-paste with an explicit import).

**Comment by mdhaber** [comment]
The Linux Tests / Nightly CPython error looks like it is happening in other PRs (e.g. gh-13661)

File: scipy/optimize/optimize.py (hunk 2 above)
Comment (Chachay) [reply scipy/optimize/optimize.py]: Thank you for your feedback.

File: scipy/optimize/optimize.py (hunk 1 above)
Comment (Chachay) [reply scipy/optimize/optimize.py:1108]: Could you elaborate what I can do for this failure?
(I'm sorry about my limit knowledge of doctest and if I'm not following some contributor guideline)

I'm a bit confused because I got the following varieties of outputs on
Windows10, Conda, Python 3.8.5 amd64, numpy 1.20.1, scipy 1.6.2&1.6.3 
```
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: 0.000000
         Iterations: 64
         Function evaluations: 330
         Gradient evaluations: 106
array([0.99999669, 0.99999329])
```

Win10, WSL2 Ubuntu20.04, Python 3.8.5, numpy 1.20.3, scipy 1.6.3 
```
Optimization terminated successfully.
         Current function value: 0.000000
         Iterations: 66
         Function evaluations: 246
         Gradient evaluations: 82
array([0.99999551, 0.99999101])
````

[Probably outdated] Win10, WSL2 Ubuntu18.04, Python 3.6.5, numpy 1.18.1, scipy 1.5.4 
```
Optimization terminated successfully.
         Current function value: 0.000000
         Iterations: 66
         Function evaluations: 328
         Gradient evaluations: 82
array([0.99999551, 0.99999101])
````

Comment (Chachay) [reply scipy/optimize/optimize.py]: Updated the output to Ubuntu 20.04 result with marked "# may vary"

File: scipy/optimize/optimize.py [hunk 3]
```diff
@@ -1088,6 +1088,40 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
     allvecs  :  list
         The value of xopt at each iteration. Only returned if retall is True.
 
+    Examples
+    --------
+    >>> from scipy.optimize import fmin_bfgs
+
```
Comment (andyfaff) [inline scipy/optimize/optimize.py:1094]: this line isn't necessary.

File: scipy/optimize/optimize.py [hunk 4]
```diff
@@ -1088,6 +1088,40 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
     allvecs  :  list
         The value of xopt at each iteration. Only returned if retall is True.
 
+    Examples
+    --------
+    >>> from scipy.optimize import fmin_bfgs
```
Comment (andyfaff) [inline scipy/optimize/optimize.py:1093]: import the rosen, rosen_grad function as well.

File: scipy/optimize/optimize.py [hunk 5]
```diff
@@ -1088,6 +1088,40 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
...
+
+    >>> def rosenbrock(x):
```
Comment (andyfaff) [inline scipy/optimize/optimize.py:1095]: Could you remove the inline version of rosenbrock? The use of minimize over the `fmin_*` methods is probably preferred, and so the full function is probably not needed.

File: scipy/optimize/optimize.py [hunk 6]
```diff
@@ -1088,6 +1088,40 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=Inf,
...
+    >>> x0 = np.array([-3, -4])
+    >>> fmin_bfgs(rosenbrock, x0)
+    Optimization terminated successfully.
+         Current function value: 0.000000
+         Iterations: 66                    # may vary
+         Function evaluations: 246         # may vary
+         Gradient evaluations: 82          # may vary
+    array([0.99999551, 0.99999101])        # may vary
+
+    >>> def rosenbrock_grad(x):
```
Comment (andyfaff) [inline scipy/optimize/optimize.py:1110]: same as above, just use the optimize.rosen_grad function.

File: scipy/optimize/optimize.py (hunk 5 above)
Comment (ev-br) [reply scipy/optimize/optimize.py:1095]: Drive-by 2c comment: a spelled-out 2D rosenbrock function would be very helpful for me as a user; way more so than an imported version. The latter does not help answering "how do I define a function to be minimized".

Comment (Chachay) [reply scipy/optimize/optimize.py:1095]: Yes, this redundant definition is from my experience when I was new to use fmin_*. Looking back at the time, it would be helpful if an example includes how to use 'args'. 

If the duplicated definition of the Rosebrock function is the matter, what about replacing this part as follow?

```
def quadratic_objective(x, Q):
    return x.T@Q@x
fmin_bfgs(quadratic_objective, x0, args=(np.diag([1., 10.]),))
```

This totally depends on my personal experience, so if this is not the case for other people, I just rewrite this to a predefined Rosenbrock version.

Comment (ev-br) [reply scipy/optimize/optimize.py:1095]: One last nit: if you go with a quadratic objective example, separate the array definition into a statement of its own. Otherwise the tuple syntax with a trailing comma is hard to parse (and this is a common source of issues for people)

Comment (Chachay) [reply scipy/optimize/optimize.py]: Updated the example. I noted the trailing comma though this might be verbose.

Comment (ev-br) [reply scipy/optimize/optimize.py]: Looks good to me now. (I _think_ a single #may vary would be enough.) 
I'll defer to @andyfaff for the decision

**Review by rgommers — APPROVED** [review]
Thanks @Chachay. I made some small formatting fixes, this looks good to go now.
